Pass the siteSlug or siteId to further steps in the hero flow to support refreshing

roo2 opened the pull request:
#### Changes proposed in this Pull Request

Fixes https://github.com/Automattic/wp-calypso/issues/58339 By passing the `siteSlug` or `siteId` which is a required dependency for all steps of the hero flow.

Also passes the query param when users use the back button.

#### Testing instructions
Navigate to 
`/start/setup-site/intent?flags=signup/hero-flow&siteSlug=YOUR_SITE.wordpress.com` or
`/start/setup-site/intent?flags=signup/hero-flow&siteId=$YOUR_SITE_ID`

Navigate through the hero flow and assert that the siteId or siteSlug is passed to the next step.
Use browser refresh to refresh the page and verify that it loads correctly.

roo2 committed:
pass the siteSlug or siteId around to further steps in the hero flow

roo2 committed:
support passing siteSlug or siteId with back button

roo2 committed:
update usages of getStepUrl to pass in a query object. Also fix importer link from intent step when a siteId is used.

arthur791004 commented on client/signup/utils.js:
```diff
@@ -1,9 +1,9 @@
+import { addQueryArgs } from 'calypso/lib/url';
 import flows from 'calypso/signup/config/flows';
 import steps from 'calypso/signup/config/steps-pure';
-
```
Still prefer to keep empty line after import section

arthur791004 commented on client/signup/main.jsx:
```diff
@@ -542,7 +542,9 @@ class Signup extends Component {
 			const locale = ! this.props.isLoggedIn ? this.props.locale : '';
-			page( getStepUrl( flowName, stepName, stepSectionName, locale ) );
+			const queryObject = ( this.props.initialContext && this.props.initialContext.query ) || {};
+			const siteSlugOrId = pick( queryObject, [ 'siteSlug', 'siteId' ] );
```
Not important but `siteIdOrSlug` might be better as we use this term everywhere in our codebase except one.

arthur791004 commented on client/signup/steps/design-picker/index.jsx:
```diff
@@ -308,6 +310,7 @@ export default function DesignPickerStep( props ) {
 			stepContent={ renderDesignPicker() }
 			align={ isReskinned ? 'left' : 'center' }
 			skipButtonAlign={ isReskinned ? 'top' : 'bottom' }
+			queryParams={ siteSlugOrId }
```
I think we can pass `queryParams` in the following place and we don't need to handle this in every step component
```js
// main.jsx
<CurrentComponent
  ...
  queryParams={ siteIdOrSlug }
/>
```

arthur791004 reviewed (commented):
Great job! How about doing it inside the `getStepUrl` function directly (I think we can get the query string from that function directly)? Then, we don't need to change so many places. But one thing that needs to take care of might be `getBackUrl` in the `NavigationLink` as it concatenates the querying with the return value of `getStepUrl`.

roo2 committed:
address PR feedback

roo2 committed:
ensure getStepUrl is being called with correctly ordered params

roo2 commented:
thanks for the feedback @arthur791004! I updated it to pass the value of `siteIdOrSlug` in via main.jsx as you suggested and this simplified the code quite a bit.

I also updated other usages of `getStepUrl` to pass in a `params` object rather than concatenating the params on the end.

I'm not sure how we should get the `siteIdOrSlug` from `getStepUrl`, we could use `window.location.search` but I think it is nicer to pass in the `params` argument because it is a clearer flow of data, even if it results in more code.

arthur791004 commented:
> I'm not sure how we should get the siteIdOrSlug from getStepUrl, we could use window.location.search but I think it is nicer to pass in the params argument because it is a clearer flow of data, even if it results in more code.

Yeah, we could use `window.location.search` to forward the query string to the previous/next step and we can avoid missing `siteIdOrSlug` in the future. But I'm okay with the current way 👍

tyxla commented on client/signup/main.jsx:
```diff
@@ -542,7 +542,9 @@ class Signup extends Component {
 		if ( stepName && ! this.isEveryStepSubmitted() ) {
 			const locale = ! this.props.isLoggedIn ? this.props.locale : '';
-			page( getStepUrl( flowName, stepName, stepSectionName, locale ) );
+			const queryObject = this.props.initialContext && this.props.initialContext.query;
+			const siteIdOrSlug = pick( queryObject, [ 'siteSlug', 'siteId' ] );
```
`pick()` is one of those Lodash functions that is very easy to replace with 1 or 2 lines of native code. 

In those instances, I'd strongly recommend not introducing an extra helper function, and using native code instead. If it's necessary in a component, you can always abstract it to a separate function.

For example, see how the usage here even makes the code simpler:

```suggestion
			const { siteSlug, siteId } = this.props.initialContext && this.props.initialContext.query;
			const siteIdOrSlug = { siteSlug, siteId };
```

roo2 replied:
oh yea... brain fart... oh well, I had fun implementing it 😂

arthur791004 replied:
@tyxla We've replaced pick with native code.

tyxla reviewed (changes requested):
I'd advise against introducing a `pick()` helper function, since it's a pretty simple functionality that can be achieved with very little code (just a simple object destructuring + creating a new object with the destructured props).

arthur791004 committed:
Replace pick with native code

arthur791004 committed:
Fix tests

tyxla commented on packages/js-utils/src/pick.ts:
```diff
@@ -0,0 +1,16 @@
+ * @param props The properties to pick
+ * @returns An object with only the "picked" properties
+ */
+export default function pick< T, U extends keyof T >( object: T, props: Array< U > ): Pick< T, U > {
```
Seems like we can delete this one from the PR now. Thanks for simplifying the logic 👍

arthur791004 replied:
I think we can keep this in case somebody wants to use it in the future? Or we can also re-implement this function at that time 🤔

tyxla replied:
The logic is very easy and implementable inline. That's preferable IMHO, rather than adding additional dependencies for one-liner functions. Not to mention that there's no reason to be introducing unused code. I suggest removing it.

arthur791004 replied:
Okay, I've removed it

arthur791004 replied:
@tyxla Okay, I've removed it!

arthur791004 committed:
Remove pick from js-utils

simison commented:
@tyxla are there/should there be Eslint warnings against using Lodash `pick()`?

tyxla commented:
> @tyxla are there/should there be Eslint warnings against using Lodash `pick()`?

Yes, but our strategy to enable them for a particular Lodash function is to do it after we've fully refactored away from the Lodash version across the codebase, in order to avoid introducing new eslint warnings.
